atc/db/team_test.go

```diff
@@ -916,6 +916,128 @@ var _ = Describe("Team", func() {
 				Expect(len(workers)).To(Equal(0))
 			})
 		})
+
+		Context("when there is a second worker", func() {
+			var atcWorker2 = atc.Worker{
+				GardenAddr:       "some-garden-addr-2",
+				BaggageclaimURL:  "some-bc-url-2",
+				HTTPProxyURL:     "some-http-proxy-url-2",
+				HTTPSProxyURL:    "some-https-proxy-url-2",
+				NoProxy:          "some-no-proxy-2",
+				ActiveContainers: 140,
+				ResourceTypes: []atc.WorkerResourceType{
+					{
+						Type:    "some-base-resource-type",
+						Image:   "some-image",
+						Version: "some-version",
+					},
+				},
+				Platform:  "some-platform",
+				Name:      "some-worker-name-2",
+				StartTime: 55,
+			}
+
+			var volumeOnWorker2 db.CreatedVolume
+
+			BeforeEach(func() {
+				_, err := workerFactory.SaveWorker(atcWorker2, 0)
+				Expect(err).ToNot(HaveOccurred())
+
+				creatingVolume, err := volumeRepository.CreateVolume(defaultTeam.ID(), atcWorker2.Name, db.VolumeTypeResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				volumeOnWorker2, err = creatingVolume.Created()
+				Expect(err).ToNot(HaveOccurred())
+
+				err = volumeOnWorker2.InitializeStreamedResourceCache(urc, atcWorker.Name)
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			Context("when the workers are running", func() {
+				BeforeEach(func() {
+					atcWorker.State = string(db.WorkerStateRunning)
+					_, err := workerFactory.SaveWorker(atcWorker, 0)
+					Expect(err).ToNot(HaveOccurred())
+
+					atcWorker2.State = string(db.WorkerStateRunning)
+					_, err = workerFactory.SaveWorker(atcWorker2, 0)
+					Expect(err).ToNot(HaveOccurred())
+				})
+
+				It("should find both workers", func() {
+					workers, err := defaultTeam.FindWorkersForResourceCache(urc.ID())
+					Expect(err).ToNot(HaveOccurred())
+					Expect(len(workers)).To(Equal(2))
+					Expect(workers[0].Name()).To(Equal("some-worker-name"))
+					Expect(workers[1].Name()).To(Equal("some-worker-name-2"))
+				})
+
+				Context("when worker1 is pruned", func() {
+					BeforeEach(func() {
+						_, err := psql.Delete("workers").Where(sq.Eq{"name": "some-worker-name"}).RunWith(dbConn).Exec()
+						Expect(err).ToNot(HaveOccurred())
+					})
+
+					It("should not find any worker", func() {
+						workers, err := defaultTeam.FindWorkersForResourceCache(urc.ID())
+						Expect(err).ToNot(HaveOccurred())
+						Expect(len(workers)).To(Equal(0))
+					})
+
+					It("cached volume should still exist on worker2", func() {
+						_, found, err := volumeRepository.FindResourceCacheVolume("some-worker-name-2", urc)
+						Expect(err).ToNot(HaveOccurred())
+						Expect(found).To(BeTrue())
+					})
+
+					Context("when worker3 fetch the resource again and streamed to worker2 again", func() {
```

Comment:
From what I understand, it is testing the cache volume on worker2 being able to be streamed from worker3 with a newly fetched resource. But the cached volume exists in worker2 already(which is covered in above test line 987). Does it matter if it can stream from worker3?

Actually, where is the test for scenario: cached volume on worker2 should not be streamed to other workers if worker1 got pruned?

Could you elaborate? Thx.

Reply:
> But the cached volume exists in worker2 already(which is covered in above test line 987). Does it matter if it can stream from worker3?
>
> Actually, where is the test for scenario: cached volume on worker2 should not be streamed to other workers if worker1 got pruned?

Line 981 "should not find any worker" verifies that, when worker-1 (source worker) is pruned, streamed cache on worker-2 will no longer be found, thus it can no longer to streamed again.

Also, worker-3 case of line 993 verifies that in the other way: after streamed the cache from worker-3 again, worker-2 holds the cache again and worker-2 now can provide the cache again.